Find and fix race condition during post-phase checks.

#36 patches over a race condition where the state disappears and you get a null pointer exception.

I haven't looked into possible causes yet.

```
[INFO ] 2022-07-20 16:01:15,473 [main]  com.oltpbenchmark.DBWorkload main - ======================================================================                              
                                                                                                                                            
Benchmark:     NOOP {com.oltpbenchmark.benchmarks.noop.NoOpBenchmark}
Configuration: /home/protect_the_innocent/noop_postgres.xml           
Type:          POSTGRES                                                                                                                                                                                                                                                                  
Driver:        org.postgresql.Driver                                                                                                        
URL:           jdbc:postgresql:/protect_the_innocent:5432/benchbase?sslmode=disable&ApplicationName=tpcc&reWriteBatchedInserts=true&prepareThreshold=0
Isolation:     TRANSACTION_READ_COMMITTED                 
Batch Size:    128                                                                                                                                                                                                                                                                       
Scale Factor:  1.0                                                                                                                          
Terminals:     200                                                    
                                                                      
[INFO ] 2022-07-20 16:01:15,474 [main]  com.oltpbenchmark.DBWorkload main - ======================================================================                              
[INFO ] 2022-07-20 16:01:15,488 [main]  com.oltpbenchmark.DBWorkload main - Creating new NOOP database...  
[INFO ] 2022-07-20 16:01:15,606 [main]  com.oltpbenchmark.DBWorkload main - Finished creating new NOOP database...
[INFO ] 2022-07-20 16:01:15,637 [main]  com.oltpbenchmark.DBWorkload main - Loading data into NOOP database...
[INFO ] 2022-07-20 16:01:15,638 [main]  com.oltpbenchmark.util.ThreadUtil runLoaderThreads - Creating a Thread Pool with a size of 1 to run 0 Loader Threads.  0 threads will be queued.
[INFO ] 2022-07-20 16:01:15,639 [main]  com.oltpbenchmark.util.ThreadUtil runLoaderThreads - Finished executing 0 Loader Threads [time=0.00s]
[INFO ] 2022-07-20 16:01:15,639 [main]  com.oltpbenchmark.DBWorkload main - Finished loading data into NOOP database...
[INFO ] 2022-07-20 16:01:15,639 [main]  com.oltpbenchmark.DBWorkload runWorkload - Creating 200 virtual terminals...
[INFO ] 2022-07-20 16:01:16,435 [main]  com.oltpbenchmark.DBWorkload runWorkload - Launching the NOOP Benchmark with 1 Phase...                                                                                                                                                          
[INFO ] 2022-07-20 16:01:16,510 [main]  com.oltpbenchmark.ThreadBench runRateLimitedMultiPhase - PHASE START :: [Workload=NOOP] [Serial=false] [Time=60] [WarmupTime=0] [Rate=unlimited] [Arrival=REGULAR] [Ratios=[100.0]] [ActiveWorkers=200]
[INFO ] 2022-07-20 16:01:17,511 [main]  com.oltpbenchmark.ThreadBench runRateLimitedMultiPhase - MEASURE :: Warmup complete, starting measurements.
[INFO ] 2022-07-20 16:02:18,510 [main]  com.oltpbenchmark.ThreadBench runRateLimitedMultiPhase - TERMINATE :: Waiting for all terminals to finish ..
[WARN ] 2022-07-20 16:02:18,511 [NoOpWorker<023>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...                                                                                                                                                                 
[WARN ] 2022-07-20 16:02:18,512 [NoOpWorker<143>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[WARN ] 2022-07-20 16:02:18,512 [NoOpWorker<150>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[WARN ] 2022-07-20 16:02:18,511 [NoOpWorker<190>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[WARN ] 2022-07-20 16:02:18,513 [NoOpWorker<159>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...                                                                                                                                                                 
[WARN ] 2022-07-20 16:02:18,513 [NoOpWorker<127>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[WARN ] 2022-07-20 16:02:18,512 [NoOpWorker<004>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[WARN ] 2022-07-20 16:02:18,512 [NoOpWorker<117>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[WARN ] 2022-07-20 16:02:18,512 [NoOpWorker<083>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...                                                                                                                                                                 
[WARN ] 2022-07-20 16:02:18,512 [NoOpWorker<007>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[WARN ] 2022-07-20 16:02:18,512 [NoOpWorker<183>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[WARN ] 2022-07-20 16:02:18,512 [NoOpWorker<003>]  com.oltpbenchmark.api.Worker run - preState is DONE? will continue...
[ERROR] 2022-07-20 16:02:18,516 [NoOpWorker<182>]  com.oltpbenchmark.ThreadBench uncaughtException - Cannot invoke "com.oltpbenchmark.Phase.getId()" because "postPhase" is null
java.lang.NullPointerException: Cannot invoke "com.oltpbenchmark.Phase.getId()" because "postPhase" is null
        at com.oltpbenchmark.api.Worker.run(Worker.java:296)
        at java.base/java.lang.Thread.run(Thread.java:833)
[ERROR] 2022-07-20 16:02:18,516 [NoOpWorker<010>]  com.oltpbenchmark.ThreadBench uncaughtException - Cannot invoke "com.oltpbenchmark.Phase.getId()" because "postPhase" is null
java.lang.NullPointerException: Cannot invoke "com.oltpbenchmark.Phase.getId()" because "postPhase" is null
        at com.oltpbenchmark.api.Worker.run(Worker.java:296)
        at java.base/java.lang.Thread.run(Thread.java:833)
```

--- a/src/main/java/com/oltpbenchmark/api/Worker.java
+++ b/src/main/java/com/oltpbenchmark/api/Worker.java
@@ -378,7 +378,7 @@ public abstract class Worker<T extends BenchmarkModule> implements Runnable {
      * implementing worker should return the TransactionType handle that was
      * executed.
      *
-     * @param databaseType    TODO
+     * @param databaseType TODO
      * @param transactionType TODO
      */
     protected final void doWork(DatabaseType databaseType, TransactionType transactionType) {

--- a/src/main/java/com/oltpbenchmark/api/Worker.java
+++ b/src/main/java/com/oltpbenchmark/api/Worker.java
@@ -296,7 +296,7 @@ public abstract class Worker<T extends BenchmarkModule> implements Runnable {
 
                         if (postPhase == null) {
                             // Need a null check on postPhase since current phase being null is used in WorkloadState
-                            // and BenchmarkState as the indication that the benchmark is over. However, there's a race
+                            // and ThreadBench as the indication that the benchmark is over. However, there's a race
                             // condition with postState not being changed from MEASURE to DONE yet, so we entered the
                             // switch. In this scenario, just break from the switch.
                             break;

--- a/src/main/java/com/oltpbenchmark/api/Worker.java
+++ b/src/main/java/com/oltpbenchmark/api/Worker.java
@@ -293,6 +293,14 @@ public abstract class Worker<T extends BenchmarkModule> implements Runnable {
                         // that either started during the warmup phase or ended
                         // after the timer went off.
                         Phase postPhase = workloadState.getCurrentPhase();
+
+                        if (postPhase == null) {
+                            // Need a null check on postPhase since current phase being null is used in WorkloadState
+                            // and BenchmarkState as the indication that the benchmark is over. However, there's a race
+                            // condition with postState not being changes to DONE from MEASURE yet, so we entered the
+                            // switch. In this scenario, just break from the switch.
+                            break;
+                        }
                         if (preState == MEASURE && postPhase.getId() == prePhase.getId()) {
                             latencies.addLatency(transactionType.getId(), start, end, this.id, prePhase.getId());
                             intervalRequests.incrementAndGet();
@@ -370,7 +378,7 @@ public abstract class Worker<T extends BenchmarkModule> implements Runnable {
      * implementing worker should return the TransactionType handle that was
      * executed.
      *
-     * @param databaseType TODO
+     * @param databaseType    TODO
      * @param transactionType TODO
      */
     protected final void doWork(DatabaseType databaseType, TransactionType transactionType) {

--- a/src/main/java/com/oltpbenchmark/api/Worker.java
+++ b/src/main/java/com/oltpbenchmark/api/Worker.java
@@ -297,7 +297,7 @@ public abstract class Worker<T extends BenchmarkModule> implements Runnable {
                         if (postPhase == null) {
                             // Need a null check on postPhase since current phase being null is used in WorkloadState
                             // and BenchmarkState as the indication that the benchmark is over. However, there's a race
-                            // condition with postState not being changes to DONE from MEASURE yet, so we entered the
+                            // condition with postState not being changed from MEASURE to DONE yet, so we entered the
                             // switch. In this scenario, just break from the switch.
                             break;
                         }

